src/mbgl/renderer/layers/render_location_indicator_layer.cpp

```diff
@@ -894,12 +1032,271 @@ void RenderLocationIndicatorLayer::render(PaintParameters& paintParameters) {
+                drawable->setIndexData(indices, std::move(drawSegments));
+
+                renderImpl->circleDrawableInfo.outlineDrawable.emplace(*drawable);
+
+                localLayerGroup->addDrawable(std::move(drawable));
+                ++stats.drawablesAdded;
+            }
+        };
+
+        createCircleDrawable();
+        createQuadDrawable(renderImpl->shadowDrawableInfo, "locationShadow");
+        createQuadDrawable(renderImpl->puckDrawableInfo, "locationPuck");
+        createQuadDrawable(renderImpl->hatDrawableInfo, "locationPuckHat");
+    };
+
+    const auto updateCircleDrawable = [&]() {
+        auto& circleDrawable = renderImpl->circleDrawableInfo.drawable.value().get();
+        auto& circleOutlineDrawable = renderImpl->circleDrawableInfo.outlineDrawable.value().get();
+
+        if (!(renderImpl->parameters.errorRadiusMeters > 0.0) ||
+            (renderImpl->parameters.errorRadiusColor.a == 0.0 &&
+             renderImpl->parameters.errorRadiusBorderColor.a == 0.0)) {
+            circleDrawable.setEnabled(false);
+            circleOutlineDrawable.setEnabled(false);
+            return;
+        } else {
+            circleDrawable.setEnabled(true);
+            circleOutlineDrawable.setEnabled(true);
+        }
+
+        // update attributes
+        if (renderImpl->circleDrawableInfo.dirty) {
+            // vertex attribute data is shared between the 2 circle drawables
+            const auto& geometry = renderImpl->circleDrawableInfo.geometry;
+
+            using VertexVector = gfx::VertexVector<RenderLocationIndicatorImpl::vec2>;
+            std::shared_ptr<VertexVector> verts = std::make_shared<VertexVector>();
+            for (const auto& elem : geometry) {
+                verts->emplace_back(elem);
+            }
+
+            auto& circleVertexAttrs = circleDrawable.getVertexAttributes();
+            if (const auto& attr = circleVertexAttrs->set(shaders::idCommonPosVertexAttribute)) {
+                attr->setSharedRawData(
+                    verts, 0, 0, sizeof(RenderLocationIndicatorImpl::vec2), gfx::AttributeDataType::Float2);
+            }
+        }
+
+        // update uniforms
+        shaders::CommonUBO commonUBO = {/* .matrix */ util::cast<float>(renderImpl->getProjectionCircle()),
+                                        /* .color */ renderImpl->parameters.errorRadiusColor};
+
+        auto& drawableUniforms = circleDrawable.mutableUniformBuffers();
+        drawableUniforms.createOrUpdate(shaders::idCommonUBO, &commonUBO, context);
+
+        commonUBO.color = renderImpl->parameters.errorRadiusBorderColor;
+        auto& outlineDrawableUniforms = circleOutlineDrawable.mutableUniformBuffers();
+        outlineDrawableUniforms.createOrUpdate(shaders::idCommonUBO, &commonUBO, context);
+    };
+
+    const shaders::CommonUBO ubo = {/* .matrix */ util::cast<float>(renderImpl->getProjectionPuck()),
+                                    /* .color */ Color::black()};
```

Comment:
This part that updates the UBO should be in the layer tweaker files: `location_indicator_layer_tweaker.hpp/cpp`.